### PR TITLE
add isUnion to macros

### LINKED ISSUE #7665: A stable way to find out if an object is a union.
In generic code it is often important to know if a type is a union or not. There should be a way to statically test for `isUnion`. This is important for the generic `$` proc in system, because it cases weird results and might even crash on union types in the current implementation.

### PR DESCRIPTION
**Pull request by krux02**
fixes #7665

### CODE REVIEW

**Comment by Araq** [comment]
Why not use the VM's callback mechanism?

**Comment by krux02** [comment]
I actually looked into the callback mechanism, but I could not find an example that uses a ``NimNode`` argument. Then I went with pattern that I know how to implement.

Btw I just realized that the implementation currently would not allow the fix for ``system.`$` ``, because it is in macros. I need to do something about that.

File: compiler/vm.nim [hunk 1]
```diff
@@ -1619,6 +1619,16 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
         regs[ra].intVal = regs[rb].node.id
       else:
         regs[ra].intVal = -1
+    of opcNIsUnion:
+      decodeB(rkInt)
+      let a = regs[rb].node
+      if a == nil or a.typ == nil:
+        stackTrace(c, tos, pc, "node has no type")
+      let typ = skipTypes(a.typ, {tyTypeDesc, tyAlias, tyGenericInst});
+      if typ.kind == tyObject and tfUnion in typ.flags:
```
Comment (timotheecour) [inline compiler/vm.nim:1628]: regs[ra].intVal = ord(tyObject and tfUnion in typ.flags)

File: lib/core/macros.nim [hunk 2]
```diff
@@ -1613,6 +1613,15 @@ macro getCustomPragmaVal*(n: typed, cp: typed{nkSym}): untyped =
   if result.kind == nnkEmpty:
     error(n.repr & " doesn't have a pragma named " & cp.repr()) # returning an empty node results in most cases in a cryptic error,
 
+when defined(nimMacrosIsUnion):
+  proc isUnion*(arg: NimNode): bool {.magic: "NIsUnion", noSideEffect, since: (1,1).}
```
Comment (timotheecour) [inline lib/core/macros.nim:1617]: * seems like a much better fit in typetraits.nim
* and instead the API should be:
```nim
proc isUnion*(t: typedesc): bool {.magic: "TypeTrait".}
```
(no need for `macro isUnion*(arg: typed)` plus `proc isUnion*(arg: NimNode): bool`)

File: lib/core/macros.nim [hunk 3]
```diff
@@ -1613,6 +1613,15 @@ macro getCustomPragmaVal*(n: typed, cp: typed{nkSym}): untyped =
...
+    ## Expects a type symbol or expression as argument. Returns
+    ## ``true`` if the underlying type is a union type. Raises, when
```
Comment (timotheecour) [inline lib/core/macros.nim:1619]: maybe give a link in doc comment to https://nim-lang.github.io/Nim/manual.html#foreign-function-interface-union-pragma

**Pull request closed without merging**